### DISCUSSION TITLE
Pusher not working in production on Vercel

### DISCUSSION THREAD

**Discussion by Hamzi-SE** [vercel/next.js#48433] [Help]
### Summary

https://nextchatter.vercel.app/

This is the link of the deployed app. In development on my local machine everything is working fine but when I deploy it on Vercel, the messages and notifications are not working until I refresh the page or send another message. (requests are delayed by one request)

This is the link to my repository:
https://github.com/Hamzi-SE/NextChatter

### Example

https://nextchatter.vercel.app/

**Comment by lightzane** [comment]
I am also encountering this issue, after deploying to Vercel, all the pusher events seems to be delayed by one request.
When sending a request, it will not be received.
After sending a 2nd request, the 1st request will show to the receiver.
When sending a 3rd request, the 2nd request will show to the receiver, and so on.

**Reply by Hamzi-SE** [reply]
Yeah encountering the same issue

**Reply by Hamzi-SE** [reply]
Found any solution? @lightzane

**Comment by NotAndreh** [comment]
I also had this issue and the solution for me was awaiting the pusher trigger function. I think that since vercel once you send the response to the client closes the serverless function, the pusher event doesn't have time to be sent

**Reply by Hamzi-SE** [reply]
Thanks for your answer @NotAndreh Can you provide the code how you did it?

**Reply by NotAndreh** [reply]
Sure, i changed 
```js
pusher.trigger('my-channel', 'my-event', {
  "message": "hello world"
});
```
to
```js
await pusher.trigger('my-channel', 'my-event', {
  "message": "hello world"
});
```
This way the serverless function does not end before pusher has sent the message

**Reply by Hamzi-SE** [reply]
Great! Worked like charm! Thanks brother ❤

**Reply by lightzane** [reply]
Awesome! Thank you @NotAndreh !

**Reply by Minna937** [reply]
Thank you!!!!!!!!!!!❤❤

**Reply by Msantanna10** [reply]
Unbelievable... I hate those kind of simple things that make you spend hours/days/weeks... Finally, I knew it was simple but I just didnt know what it was. Thank you!

**Reply by dlgudwn123** [reply]
but isn't it is too slow?

**Reply by OmidHajizadeh** [reply]
I HAVE BEEN BANGING MY HEAD OVER THIS FOR DAYS!!
THANK YOU SIR!!!!

**Reply by scottbedard** [reply]
Fantastic solution and explanation, thank you. This also fixes the problem in a Netlify environment.

**Reply by Yagth** [reply]
This was really helpful. Thanks man

**Reply by PradneyaSP** [reply]
You are a lifesaver mann !!

**Reply by codevandennis** [reply]
Thank you! This will save lives...

**Reply by alexgoni** [reply]
Thank you so much!!!!

**Reply by reibal** [reply]
@NotAndreh Man, you're the MVP! Thank you so much!! I was going crazy xD now it works like a charm, I can't thank you enough!

**Reply by petrinichpavel** [reply]
Thank you!!!!

**Reply by Lenic** [reply]
Thank you, 🙏. You have solved the problem that has been bothering me all night.

**Reply by mmagnusen** [reply]
THANK YOU!!! I have been scratching my head over this for hours 😅

**Comment by PiotrBryla** [comment]
It's also worth remembering that Vercel will cache your `GET` requests by default, so when using Pusher triggers, use a different HTTP method.

**Reply by manuelffernandez** [reply]
I'm not really sure about this, but I think it's not necessary to replace the `GET` convention method to skip the cache behaviour. From what I saw in [NextJS docs](https://nextjs.org/docs/app/building-your-application/routing/route-handlers#opting-out-of-caching) it's possible to avoid caching by exporting the route segment config 'dynamic' with 'force-dynamic' value

```javascript
export const dynamic = 'force-dynamic'

export default GET function() {
  // Your code...
}
```

**Comment by nicitaacom** [comment]
@manuelffernandez You are right - thank you for your answer you helped to solve this issue - https://streamable.com/dze31q
(I added `export const dynamic = 'force-dynamic'` to `/support/tickets/[ticketId]/page.tsx` and `MessagesBody.tsx`)
@NotAndreh - thank you I just added `await` and now pusher work in production as well

Update: ok sometimes it doesn't fix this https://streamable.com/dze31q (I added router.refresh() on DesktopSidebarTicekt click - still the same)

### Update 2: I fixed it in production - so your function to open some chat/ticket should look like this
```ts
 function openTicket(ticketId: string) {
    resetUnreadMessages(ticketId)
    setTimeout(() => {
      router.refresh()
    }, 250)
  }
  ```
  
I added timeout and it fiexs issue shown here https://streamable.com/dze31q

<br/>
<br/>
<br/>

### Possible fix 1: add `await pusherServer.trigger` something

### Possible fix 2: add `force-dynamic`
```ts

// This is required to show all tickets in prodcuton
export const dynamic = "force-dynamic"
```

For me whole code looks like this

**app/(support)/support/tickets/layout.tsx**
```tsx

export const metadata: Metadata = {
  title: "Hot Delivery - support",
  description: "Support on hot-delivery.vercel.app - we deliver you order as fresh as possible",
}

// This is required to show all tickets in prodcuton
export const dynamic = "force-dynamic"

export default async function SupportChatLayout({ children }: { children: React.ReactNode }) {
  const initialTickets = await getInitialTickets()
  const unreadMessages = await getUnreadMessages()

  // it is protected route and only SUPPORT role has access to this route
  const { data: role_response, error: anonymous_user } = await supabaseServer().from("users").select("role").single()
  // Allow SUPPORT to visit this page
  if (!role_response?.role.includes("SUPPORT") || anonymous_user) {
    redirect("/")
  }

  return (
    <>
      <Navbar />
      <div
        className="w-full h-[calc(100vh-81px)] overflow-x-hidden overflow-y-auto hide-scrollbar
      flex flex-row">
        <DesktopSidebar unseenMessages={unreadMessages ?? []} initialTickets={initialTickets} />
        <MobileSidebar unseenMessages={unreadMessages ?? []} initialTickets={initialTickets} />
        {children}
      </div>
    </>
  )
}
```
